### PR TITLE
[mds-compliance] Feature/jane/refactor compliance engine for readability, and add time violations to total_violations

### PR DESCRIPTION
**Pull request by janedotx**
## PR Checklist

 - [ ] simple searchable title - `[mds-db] Add PG env var`, `[config] Fix eslint config`
 - [ ] briefly describe the changes in this PR
 - [ ] mark as draft if should not be merged
 - [ ] write tests for all new functionality

## Impacts
- [ ] Provider
- [ ] Agency
- [ ] Audit
- [ ] Policy
- [x] Compliance
- [ ] Daily
- [ ] Native
- [ ] Policy Author

### CODE REVIEW

File: packages/mds-compliance/mds-compliance-engine.ts [hunk 1]
```diff
@@ -43,6 +43,7 @@ import moment from 'moment-timezone'
 const { env } = process
 
 const TWO_DAYS_IN_MS = 172800000
+type MatchedVehiclePlusRule = MatchedVehicle & { rule_id: UUID }
```
Comment (avatarneil) [inline packages/mds-compliance/mds-compliance-engine.ts]: We have a `types.ts` file for `mds-compliance` this should probably live in.

File: packages/mds-compliance/mds-compliance-engine.ts [hunk 2]
```diff
@@ -281,39 +273,36 @@ function processPolicy(
             return [...acc, ...map.matched]
           }, [])
 
-          const overflowVehicles = bucketMap.reduce((acc: (MatchedVehicle & { rule_id: UUID })[], map) => {
-            return [...acc, ...map.overflowed]
-          }, [])
+          bucketMap.forEach(bucket => {
```
Comment (avatarneil) [inline packages/mds-compliance/mds-compliance-engine.ts]: This is the textbook definition of what a `reduce` function should be doing, I don't like this change.

File: packages/mds-compliance/mds-compliance-engine.ts [hunk 3]
```diff
@@ -281,39 +273,36 @@ function processPolicy(
...
-                device: MatchedVehicle & { rule_id: UUID }
-              ) => {
-                acc[device.device_id] = device
-                return acc
-              },
-              {}
-            )
-          }
+
+          allOverflowVehicles.forEach(vehicle => {
```
Comment (avatarneil) [inline packages/mds-compliance/mds-compliance-engine.ts]: ibid, this should be a reduction

**Review by avatarneil — CHANGES_REQUESTED** [review]

File: packages/mds-compliance/mds-compliance-engine.ts (hunk 3 above)
Comment (janedotx) [reply packages/mds-compliance/mds-compliance-engine.ts]: okay

File: packages/mds-compliance/mds-compliance-engine.ts [hunk 4]
```diff
@@ -353,8 +342,8 @@ function processPolicy(
     return {
       policy,
       compliance,
-      total_violations: overflowedVehicles.length + allSpeedingVehicles.length,
-      vehicles_in_violation: { ...overflowedVehicles, ...speedingVehiclesMap }
+      total_violations: overflowedVehicles.length + allTimeVehicles.length + allSpeedingVehicles.length,
+      vehicles_in_violation: { ...overflowedVehicles, ...timeVehiclesMap, ...speedingVehiclesMap }
```
Comment (janedotx) [inline packages/mds-compliance/mds-compliance-engine.ts:353]: we need better testing/better type annotations. i just realized the type of `timeVehiclesMap` and `speedingVehiclesMap` do not match `overflowedVehicles`, but tests passed regardless.

**Pull request closed without merging**